File: internal/dto/material.go
package dto

type CategoryNode struct {
	ID       uint64         `json:"id"`
	ParentID uint64         `json:"parentId"`
	Name     string         `json:"name"`
	Code     string         `json:"code"`
	Sort     int            `json:"sort"`
	Enabled  int            `json:"enabled"`
	Children []CategoryNode `json:"children,omitempty"`
}

type CategoryCreateReq struct {
	ParentID uint64 `json:"parentId"`
	Name     string `json:"name" binding:"required"`
	Sort     int    `json:"sort"`
	Enabled  *int   `json:"enabled"`
}

type CategoryUpdateReq struct {
	ParentID *uint64 `json:"parentId"`
	Name     *string `json:"name"`
	Sort     *int    `json:"sort"`
	Enabled  *int    `json:"enabled"`
}

type CategoryReorderReq struct {
	Items []CategoryReorderItem `json:"items" binding:"required"`
}

type CategoryReorderItem struct {
	ID       uint64 `json:"id" binding:"required"`
	ParentID uint64 `json:"parentId"`
	Sort     int    `json:"sort"`
}

type MaterialDTO struct {
	ID         uint64   `json:"id"`
	CategoryID uint64   `json:"categoryId"`
	Title      string   `json:"title"`
	MediaType  string   `json:"mediaType"`
	URL        string   `json:"url"`
	CoverURL   string   `json:"coverUrl"`
	FileName   string   `json:"fileName"`
	Mime       string   `json:"mime"`
	SizeBytes  int64    `json:"sizeBytes"`
	DurationMs int64    `json:"durationMs"`
	Width      int      `json:"width"`
	Height     int      `json:"height"`
	Tags       []string `json:"tags"`
	Remark     string   `json:"remark"`
	ProductID  *uint64  `json:"productId"`
	ProductSn  string   `json:"productSn"`
	Sort       int      `json:"sort"`
	Status     int      `json:"status"`
	CreatedBy  uint64   `json:"createdBy"`
	CreatedAt  string   `json:"createdAt"`
	UpdatedAt  string   `json:"updatedAt"`
}

type MaterialCreateReq struct {
	CategoryID uint64   `json:"categoryId"`
	Title      string   `json:"title" binding:"required"`
	MediaType  string   `json:"mediaType" binding:"required"`
	URL        string   `json:"url" binding:"required"`
	CoverURL   string   `json:"coverUrl"`
	FileName   string   `json:"fileName"`
	Mime       string   `json:"mime"`
	SizeBytes  int64    `json:"sizeBytes"`
	DurationMs int64    `json:"durationMs"`
	Width      int      `json:"width"`
	Height     int      `json:"height"`
	Tags       []string `json:"tags"`
	Remark     string   `json:"remark"`
	ProductID  *uint64  `json:"productId"`
	ProductSn  string   `json:"productSn"`
	Sort       int      `json:"sort"`
}

type MaterialUpdateReq struct {
	CategoryID *uint64  `json:"categoryId"`
	Title      *string  `json:"title"`
	CoverURL   *string  `json:"coverUrl"`
	Tags       []string `json:"tags"`
	Remark     *string  `json:"remark"`
	ProductID  *uint64  `json:"productId"`
	ClearProductID bool `json:"clearProductId"`
	ProductSn  *string  `json:"productSn"`
	Sort       *int     `json:"sort"`
	Status     *int     `json:"status"`
}

type MaterialListQuery struct {
	CategoryID  uint64 `form:"categoryId"`
	IncludeChildren bool `form:"includeChildren"`
	MediaType   string `form:"mediaType"`
	Keyword     string `form:"keyword"`
	ProductSn   string `form:"productSn"`
	HasProduct  string `form:"hasProduct"` // yes | no | ""
	Tag         string `form:"tag"`
	Page        int    `form:"page"`
	PageSize    int    `form:"pageSize"`
}

type BatchIDsReq struct {
	IDs []uint64 `json:"ids" binding:"required"`
}

type DashboardStats struct {
	MaterialCount    int64            `json:"materialCount"`
	ImageCount       int64            `json:"imageCount"`
	VideoCount       int64            `json:"videoCount"`
	WeekNewCount     int64            `json:"weekNewCount"`
	CategoryCount    int64            `json:"categoryCount"`
	ByCategory       []CategoryCount  `json:"byCategory"`
}

type CategoryCount struct {
	CategoryID   uint64 `json:"categoryId"`
	CategoryName string `json:"categoryName"`
	Count        int64  `json:"count"`
}

File: web/src/api/material.ts
import client, { unwrap, type PageData } from './client'

export interface CategoryNode {
  id: number
  parentId: number
  name: string
  code: string
  sort: number
  enabled: number
  children?: CategoryNode[]
}

export interface Material {
  id: number
  categoryId: number
  title: string
  mediaType: 'image' | 'video'
  url: string
  coverUrl: string
  fileName: string
  mime: string
  sizeBytes: number
  durationMs: number
  width: number
  height: number
  tags: string[]
  remark: string
  productId?: number | null
  productSn: string
  sort: number
  status: number
  createdBy: number
  createdAt: string
  updatedAt: string
}

export interface DashboardStats {
  materialCount: number
  imageCount: number
  videoCount: number
  weekNewCount: number
  categoryCount: number
  byCategory: { categoryId: number; categoryName: string; count: number }[]
}

export interface MaterialListQuery {
  categoryId?: number
  includeChildren?: boolean
  mediaType?: string
  keyword?: string
  productSn?: string
  hasProduct?: string
  tag?: string
  page?: number
  pageSize?: number
}

export async function fetchDashboardStats(): Promise<DashboardStats> {
  const res = await client.get('/dashboard/stats')
  return unwrap(res)
}

export async function listCategories(): Promise<CategoryNode[]> {
  const res = await client.get('/categories')
  return unwrap(res) || []
}

export async function createCategory(body: {
  parentId?: number
  name: string
  sort?: number
  enabled?: number
}): Promise<CategoryNode> {
  const res = await client.post('/categories', body)
  return unwrap(res)
}

export async function updateCategory(
  id: number,
  body: Partial<{ parentId: number; name: string; sort: number; enabled: number }>,
): Promise<CategoryNode> {
  const res = await client.put(`/categories/${id}`, body)
  return unwrap(res)
}

export async function deleteCategory(id: number): Promise<void> {
  await client.delete(`/categories/${id}`)
}

export async function reorderCategories(
  items: { id: number; parentId: number; sort: number }[],
): Promise<void> {
  await client.put('/categories/reorder', { items })
}

export async function listMaterials(query: MaterialListQuery): Promise<PageData<Material>> {
  const res = await client.get('/materials', { params: query })
  return unwrap(res)
}

export async function getMaterial(id: number): Promise<Material> {
  const res = await client.get(`/materials/${id}`)
  return unwrap(res)
}

export async function createMaterial(body: {
  categoryId?: number
  title: string
  mediaType: string
  url: string
  coverUrl?: string
  fileName?: string
  mime?: string
  sizeBytes?: number
  tags?: string[]
  remark?: string
  productId?: number | null
  productSn?: string
}): Promise<Material> {
  const res = await client.post('/materials', body)
  return unwrap(res)
}

export async function updateMaterial(
  id: number,
  body: Partial<{
    categoryId: number
    title: string
    coverUrl: string
    tags: string[]
    remark: string
    productId: number | null
    clearProductId: boolean
    productSn: string
    sort: number
  }>,
): Promise<Material> {
  const res = await client.put(`/materials/${id}`, body)
  return unwrap(res)
}

export async function deleteMaterial(id: number): Promise<void> {
  await client.delete(`/materials/${id}`)
}

export async function batchDeleteMaterials(ids: number[]): Promise<void> {
  await client.post('/materials/batch-delete', { ids })
}

/** flatten category tree for select options */
export function flattenCategories(
  nodes: CategoryNode[],
  depth = 0,
): { id: number; name: string; code: string; depth: number; disabled?: boolean }[] {
  const out: { id: number; name: string; code: string; depth: number }[] = []
  for (const n of nodes) {
    out.push({ id: n.id, name: n.name, code: n.code, depth })
    if (n.children?.length) {
      out.push(...flattenCategories(n.children, depth + 1))
    }
  }
  return out
}
